processor/metricstransformprocessor/metrics_transform_processor_helper.go

```diff
@@ -0,0 +1,372 @@
+}
+
+// update updates the metric content based on operations indicated in transform.
+func (mtp *metricsTransformProcessor) update(metric *metricspb.Metric, transform Transform) {
+	// metric name update
+	if transform.NewName != "" {
+		metric.MetricDescriptor.Name = transform.NewName
+	}
+
+	for _, op := range transform.Operations {
+		switch op.Action {
+		case UpdateLabel:
+			mtp.updateLabelOp(metric, op)
+		case AggregateLabels, AggregateLabelValues:
+			mtp.aggregateOp(metric, op)
+		}
+	}
+}
+
+// updateLabelOp updates labels and label values in metric based on given operation
+func (mtp *metricsTransformProcessor) updateLabelOp(metric *metricspb.Metric, op Operation) {
+	for idx, label := range metric.MetricDescriptor.LabelKeys {
+		if label.Key != op.Label {
+			continue
+		}
+		// label key update
+		if op.NewLabel != "" {
+			label.Key = op.NewLabel
+		}
+		// label value update
+		labelValuesMapping := op.ValueActionsMapping
+		for _, timeseries := range metric.Timeseries {
+			newValue, ok := labelValuesMapping[timeseries.LabelValues[idx].Value]
+			if ok {
+				timeseries.LabelValues[idx].Value = newValue
+			}
+		}
+	}
+}
+
+// aggregateOp aggregates the data points in metric based on given operation
+func (mtp *metricsTransformProcessor) aggregateOp(metric *metricspb.Metric, op Operation) {
+	labelSet := op.LabelSetMap
+	// labelIdxs is a slice containing the indices of the selected labels.
+	// This is needed because label values are ordered in the same order as the labels
+	labelIdxs, labels := mtp.getLabelIdxs(metric, labelSet)
+	timeseriesGroup := mtp.groupTimeseries(metric, labelIdxs, op.NewValue, op.AggregatedValuesSet)
+	// merge groups of timeseries
+	newTimeseries := mtp.aggregateTimeseriesGroups(timeseriesGroup.keyToTimeseriesMap, timeseriesGroup.keyToLabelValuesMap, op.AggregationType, metric.MetricDescriptor.Type)
+	metric.Timeseries = newTimeseries
+	if op.Action == AggregateLabels {
+		metric.MetricDescriptor.LabelKeys = labels
+	}
+}
+
+// getLabelIdxs gets the indices of the labelSet labels' indices in the metric's descriptor's labels field
+// Returns the indices slice and a slice of the actual labels selected by this slice of indices
+func (mtp *metricsTransformProcessor) getLabelIdxs(metric *metricspb.Metric, labelSet map[string]bool) ([]int, []*metricspb.LabelKey) {
+	labelIdxs := make([]int, 0)
+	labels := make([]*metricspb.LabelKey, 0)
+	for idx, label := range metric.MetricDescriptor.LabelKeys {
+		_, ok := labelSet[label.Key]
+		if ok {
+			labelIdxs = append(labelIdxs, idx)
+			labels = append(labels, label)
+		}
+	}
+	return labelIdxs, labels
+}
+
+// groupTimeseries groups all timeseries in the metric that will be aggregated together based on the selected labels' values indicated by labelIdxs OR on the entire label values after replacing the aggregatedValues by newValue.
+// Depending on if newValue is set, the approach to group timeseries is different.
```

Comment:
I still find this really hard to follow sorry. Is the fact that `NewValue` is set determining that we are doing an `AggregateLabelValues` operation?

Generally speaking you want to make things much more explicit than that. Both for readability and for modularity - e.g. if I've understood this correctly then it seems like this would break in a strange way if a user set the operation to `AggregateLabels` but also supplied `NewValue`?

Separating the code more clearly based on operation will help here too. But consider if there are any other cases like this that can be made more explicit.

Reply:
Great feedback. I have separated this into two functions in different files.